### ISSUE TITLE
Metadata: base keys and registry

### ISSUE THREAD

**Issue by kaniini** [ircv3/ircv3-specifications#133]
Now that we have a working METADATA stack somewhere, I think it's time to start talking about base keys and registry.  This can run without blocking 3.2 once the other issues with METADATA are fixed, just by ensuring the documentation describes the registry and where it is located etc.  Obviously the registry should probably run parallel to actual protocol revisions because it's just data and suggestions for what the data means and it's interpretation.  We are already *sort* of doing this with batch types...

## Proposed base keys

These are some initial suggestions for the registry.

| key         | direction         | semantic interpretation | example |
|-------------|--------------------|--------------------------------|-------------|
| `avatar` | *client-set*       | A URI which points to an image resource for the user, this is a hint for graphical clients which wish to display avatars for users. | `avatar :http://turtle.dereferenced.org/~kaniini/avatar.png` |
| `certfp`  | *server-set*     | The user's certificate fingerprint, if any, in modern IRC optionally replaces the legacy `RPL_WHOISCERTFP` numeric. The fingerprint is a hexadecimal digest, specified by the `certfp-hash` metakey. | `certfp :1234567890abcdef` |
| `certfp-hash` | *server-set*   | The hash used for the certificate fingerprint if it is set. | `certfp-hash :sha-2-256` |
| `info` | *client-set* | General descriptive information about the client object, for example set by a bot. | `info :This bot scans for open proxies.  Find out more at http://irc.dereferenced.org/bots` |
| `url` | *client-set* | A URL which points to a web resource concerning the user, such as a blog or homepage. | `url :http://kaniini.dereferenced.org/` |
| `version` | *client-set* | Descriptive text describing the software implementation of the connected client. | `version :irssi v0.8.18 -- running on FreeBSD amd64` |

## Best practices concerning the registry

To avoid conflicts, keys which are *not* presently documented in the registry SHOULD use a vendor namespace, using the same rules as the CAP extensions.  Clients SHOULD NOT attempt to do any special interpretation of unrecognized keys, instead displaying the raw data.

The registry should, as previously mentioned, run parallel to standards development cycles, allowing for the registry to move at an appropriate pace for client authors and networks.

**Comment by DanielOaks** [comment]
There is this line in the METADATA spec we would want to have another look at: ``Server and client authors cannot assume that keys will carry specific formatting or meaning.``

If we're going to push the registry and try to get/keep most software compliant with it, would we want to put some of that best-practices info into the spec as well and integrate it directly, make it more 'authoritative'?

If clients and servers are gonna end up treating these keys as de facto standards in regular use, specifically around keys like `account` and `away` where clients would almost have to take them to mean the above, would we want to be more explicit about these? Seems like the line from the spec above and keys being defined with those sort of meanings could contradict.

Specifically, would the recommended keys up there that are server-set ones be better defined as `properties`, which are more explicit?

**Comment by kaniini** [comment]
@DanielOaks To the client what is the difference between a `property` and metadata in this regard?  3.3's `property` objects are just `mode`s, but with names instead of letters.

**Comment by jwheare** [comment]
The original issue has been modified after some IRC discussion, but I'll leave some rationale for my thinking around this here for posterity.

In general, I still don't really favour having any sort of official but hand-wavy registry at the moment, because it makes it hard to know how much attention to pay to it. The whole thing can basically be written off as optional, which strikes me as a bad way to achieve widespread adoption in clients and servers.

That said, there are some useful metadata keys that it would be beneficial to have some sort of documented consensus on. For me, this is limited to things that currently have no widely adopted standard alternatives, either new desirable features (avatar), or features that have been implemented in a fragmented way (certfp enrollment with nickserv etc).

However, it should not include things for which we already have standard accepted consensus on, e.g. `away` and `account` (previously included in this list, but since removed)

As a client dev, adding metadata support for stuff that already has consensus via a different numeric is feasible, but an annoyance and distraction from more important work. duplicated effort for no tangible gain.

As a server dev, if I want good client compatibility I would never choose to only use away metadata, so again, no tangible benefit as I would have to still support the existing numerics, e.g. AWAY, RPL_WHOISAWAY, ACCOUNT, along with corresponding CAP away-notify account-notify etc.

So to recap, I'm still wary about the idea of a registry, but if it goes ahead, restrict it to things that are currently broken/fractured/missing from existing implementations.

**Comment by DanielOaks** [comment]
Alright, here's my take on it.

The problem with a registry is we have keys that clients' codepaths may/will depend on, so we don't want those to be settable to anything willy-nilly. So we have a few aspects of this to look at.

------

Namespaced Metadata
-------------------------------
Metadata keys may be namespaced. This is done with the `/`, so a key `away` in the `mammon` namespace would end up being: `mammon/away`.

By default, namespaced keys are not settable by users. However, servers MAY let clients set certain namespaced keys on a key-by-key basis.

Namespaced keys can be listed using `METADATA LIST`, and may be retrieved by clients using `METADATA GET` including the appropriate prefixes.

Smart Metadata
----------------------
There are some keys whose values MUST be standard, as clients' codepaths change based on them. These keys are not user-settable, except with either specific checks to make sure they comply with standard values, or (as with things like a possible `mammon/account` for the logged-in user account name), through other non-`METADATA` commands.

The `ircv3` namespace is reserved for smart keys defined in this registry. If a server exposes an `ircv3`-namespaced key, it MUST conform to the specifications for that smart key in this registry.

Non-namespaced keys in this registry SHOULD conform to these expected definitions, but server and client authors cannot assume that keys will carry specific formatting or meaning, as per regular METADATA keys.

-----

Thoughts on something along these lines?

Though, we do need to have some way of letting the client know the current METADATA namespace, perhaps a `METADATA_NAMESPACE=` or whatever-named `ISUPPORT` thing to let clients know?

**Comment by grawity** [comment]
Sounds a bit too complex... I don't like how the client needs to use different names (`mammon/account`, `ircu/account` etc.) for the same thing.

Like: If it wants to (for example) set an away message, should it always just set `{ns}/away` expecting it to always behave the same way? If so, what's the whole point of vendor namespace? And if `{ns}/away` _doesn't_ behave the same way across all vendors, that's even worse. Remember the success of HTML & CSS "vendor prefixes"…

_(IMHO, the only sane use for vendor prefixes is based on who originally specified their usage, regardless of who implements them. For example, `znc.in/foobar` meaning that the token was defined by znc.in – as opposed to `foobar` meaning IRCv3 WG spec.)_

I don't have significantly better ideas yet, though I kinda prefer namespacing based on type (similar to Linux xattrs), but don't know how that'd work exactly.

**Comment by DanielOaks** [comment]
I don't mean to put words in anyone's mouth, but @jwheare rejects the `account`/`away` keys, not sure what @kaniini thinks since they've been removed from the original post.

We were talking about putting them under the `ircv3` namespace, little log:

```
21:10:44 < jwheare> fwiw i would oppose a ircv3/away
21:10:51 < jwheare> would be happier with mammon/away etc
21:11:14 < dan-> fair enough, ircv3/account ?
21:11:15 < jwheare> but agree on ircv3/certfp ircv3/avatar etc
21:11:22 < jwheare> no to ircv3/account
```

This was when namespace/etc discussion in the channel was still a bit confused however, so some clarification may be needed.

We're probably going to try and map at least `away` to the metadata system regardless, I believe. Either under `ircv3` if it gets standardized or `mammon` if it's rejected (depending on how just the namespacing idea goes).

**Comment by jwheare** [comment]
Fwiw, I don't really understand the namespacing use case at the moment, so that chatlog is probably a bit meaningless out of context. This issue is about whether it's useful to have a registry of consensus-approved metadata keys that can be considered the beginnings of a standard. Namespacing/vendor-prefixing/smart vs user metadata are making this quite confusing to follow.

**Comment by grawity** [comment]
Either way, using the _presence_ of a vendor prefix to determine the magic level of a metadata key is a bad idea. These are compeltely separate properties.

---

I'm thinking that the keys could first be namespaced by usage…

- `user.*` for user-settable metadata without any checks, meaningless to the server (although clients may of course attach their own meanings to specific keys);

- `sys.*` for _possibly_ user-settable metadata ("smart metadata"), where the behavior is defined individually for each key (e.g. `sys.account` would require SASL, while `sys.vhost` would require a specific oper privilege);

…and then by vendor:

- `user.<vendor>/*` and `sys.<vendor>/*` when the meaning is defined by a spec published by &lt;vendor&gt;;

- `user.*` and `sys.*` when the meaning is defined by IRCv3 WG.

(Feel free to choose your own punctuation.)

**Comment by kaniini** [comment]
I think for now I am just going to withdraw this proposal, because it's suffering the same over-engineering problem as before.  If someone else values the concept then they can open another proposal.

At the end of the day what I had in mind was a registry solely so that if someone defined a metakey, others might use the same metakey for the same meaning.
